src/main/java/io/quarkus/security/PermissionChecker.java

```diff
@@ -0,0 +1,65 @@
+package io.quarkus.security;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation that can be used to annotate a CDI bean method that checks
+ * if a {@link io.quarkus.security.identity.SecurityIdentity} holds a permission specified by the {@link #value()}.
+ * For example:
+ * <pre>
+ * {@code
+ * &#64;Path("hello")
+ * public class HelloResource {
+ *
+ *     &#64;PermissionsAllowed("speak")
```

Comment:
Any of you worried about the potential conflicts with traditional permissions? That's the first thing that came to my mind when having a look at this proposal but maybe I'm being too cautious.

I'm not sure another annotation makes sense but I wonder if there should be a marker (maybe `speak()`) to make it clear that it's a method and not your regular permission.

I'm just pushing this out there to see what you think about it. Again, I might be too cautious. The specific case I have in mind is someone having a permission that somehow gets overridden when the method is created or the opposite.

Reply:
> Any of you worried about the potential conflicts with traditional permissions? 

`@PermissionsAllowed` specifies required permissions, this PR not API makes no changes in regard of this annotation.

Before this PR, you had 2 options how to authorize required permissions:

- add `SecurityIdentity` permission checker that did `possessedPermission.implies(requiredPermission)`
- map roles to permissions and Quarkus added for you ^^^
- token claims were mapped to ^^^

What this PR adds is another way to add this permission checker to the `SecurityIdentity`.

> The specific case I have in mind is someone having a permission that somehow gets overridden when the method is created or the opposite.

I made a "trick" that prohibits that. I changed build time generation of required permissions so that `@PermissionsAllowed("read")` that is matched to `@PermissionChecker("read")` doesn't have name `read`. Instead, the name is generated permissions class name. We will never be able to avoid situation when user adds itself some custom checker that says `grant access to every permissions` which would include `@PermissionsAllowed("read")` but that comes from customizable security. We cannot exclude Quarkus users completely.

Reply:
@gsmet I re-read your comment again and I am not sure I have answered your question, feel free to ask again or let me know what I misunderstood.

Reply:
Yeah I don't think you have answered my question.

My point is that it might be easy to get confused between permissions and permission checkers during the life of an application if there's no marker to differentiate them.

Again I'm not saying I'm right, I'm just pointing to a potential issue.

Reply:
> My point is that it might be easy to get confused between permissions and permission checkers during the life of an application if there's no marker to differentiate them.

Ah, sorry about that, now I get what you are saying. I don't think it is possible to entirely avoid this. It is only possible if signatures are matching because the `@PermissionChecker` must return boolean, but for example CRUD delete methods that accepts ID can have matching method signature.

In my eyes, it is user failure, but if we can do something to help users, I am open to changes.

> I'm not sure another annotation makes sense but I wonder if there should be a marker (maybe speak()) to make it clear that it's a method and not your regular permission.

If you look to my tests in the https://github.com/quarkusio/quarkus/pull/43846 you will see that you can easily combine possessed permissions and permission checkers and IMHO that how it should be. This new annotation shouldn't reduce `@PermissionsAllowed` capabilities (it's repeatable annotation, you can switch `AND/OR` with `inclusive` flag). If we start treat `@PermissionsAllowed("read")` differently just because it is checked by this `@PermissionChecker("read")`, it means we need to differentiate between them. My personal opinion is not to do that.

Reply:
Hi @gsmet 

>My point is that it might be easy to get confused between permissions and permission checkers during the life of an application if there's no marker to differentiate them.

Can you please type some code to show what you are concerned about ? Say start with code fragment 1 and then show the code fragment 2 highlighting the possible issue ?

`@PermissionAllowed("speak")` is like `@RolesAllowed("speak")`  in a sense that both only state the authorization restrictions that must be met before the method is allowed to be invoked, both are set on the secured method.

`@PermissionChecker` is about marking a CDI bean method which will be used to assert with a boolean response if the security identity possesses a `speak` permission, and in this case the fact that the identity has the required permission can be deduced by checking JAX-RS request parameters, as opposed to a more difficult mechanism which also plan to simplify.

This is essentially our way of avoiding PerAuthorize like solutions with expression languages, etc... This PR will also let users avoid having to deal directly with Java Permission classes where it can make sense...

Hope it clarifies a few things, but please type some code to make it clearer what the concern is

Reply:
@sberyozkin IIUC what @gsmet is worried about is that users will do this (metacode):

```
@Path("product")
public class ProductResource {
 
 @PermissionChecker("delete-product")
 @Path("{id}")
 @GET
  public boolean delete(@PathParam("id") Long productId, SecurityContext securityContext) {
     return Product.deleteById(productId);
  }

  @PermissionsAllowed("delete-product")
  boolean canDelete(Long productId, SecurityContext securityContext) {
       var user =  User.findByName(securityContext.getPrincipal().getName());
       var product = Product.getById(productId);
       return userOwnsProduct(user, product);
  }
}
```

so because they share signature and both returns `boolean`, they can be confused if user made mistake and didn't bother to test security. What we can do about that:

- validate that `@PermissionChecker` is not used together with app entry point, that is `ExecutionModelAnnotationsAllowedBuildItem` ?

then there are options suggested by @gsmet , IIUC they will require explicit match between `@PermissionsAllowed(value = "owner", usePermissionChecker = true)` and `@PermissionChecker("owner")`. I might have misunderstood that! Anyone, reason why this cannot be done is that `@PermissionsAllowed` allows following:

- one of `@PermissionAllowed({ "checker1", "checker2", "possessedPermission1"})`
- all of `@PermissionAllowed(value = { "checker1", "checker2", "possessedPermission1"}, inclusive = true)`

Also, it is one more thing that user needs to do. I think that if this confusion happens, user made mistake.

Reply:
Considering you need matching signature that returns boolean, is new proposed validation using `ExecutionModelAnnotationsAllowedBuildItem` sufficient response @gsmet ? Because it will only limit possible confusion to CDI beans and we do advise to prefer endpoint-validation inside docs (I think I wrote it somewhere TBH, didn't check now).

Reply:
@michalvavrik, thanks, I think users can do a lot of strange things not only with mixing up PermissionCheckers and PermissionAllowed. Besides, I don't think anyone is actually writing JAX-RS methods returning number or booleans like `true` or `1`. 

But since it is technically possible, let's tighten it a bit, but it is a Quarkus level concern, here, at the API level, we only introduce the actual concept, `PermissionChecker`.

I propose, at the Quarkus level, fail the build if a user put a PermissionChecker on the JAX-RS method, do you think it is possible to do ?

Reply:
Yes @sberyozkin it is and I'll add that validation. It is not a difficult task as `io.quarkus.resteasy.reactive.common.deployment.JaxrsMethodsProcessor#jaxrsMethods` also produces `ExecutionModelAnnotationsAllowedBuildItem`. And yeah, I agree with your comment, but it's @gsmet that needs to ACK.